tests(rust): fix flaky transport tests

adrianbenavides opened the pull request:
Transport tests that were using a random port to create the "listener" sides were failing when the selected port was already used by another process. These changes add a loop to retry binding a transport until it succeeds.

It also adds a small change to the `test` macro so that we don't have to explicitly stop the context at the end of every test.

Fixes https://github.com/build-trust/ockam/issues/2720

adrianbenavides committed:
test(rust): fix flaky transport tests

Transport tests that were using a random port to create the "listener" sides were
failing when the selected port was already used by another process. These changes
add a loop to retry binding a transport until it succeeds.

It also adds a small change to the `test` macro so that we don't have to explicitly
stop the context at the end of every test.

twittner commented:
Can we not have the operating system assign us a port number by using 0?

adrianbenavides commented:
> Can we not have the operating system assign us a port number by using 0?

We should be able to by modifying the `listen` and `create_outlet` methods on the transports crates to return the assigned address. That would definitely be an improvement. Maybe we could leave this for a separate PR? Just to ensure that the tests are stable now without changing any production code.

Thanks for the tip @twittner!

adrianbenavides committed:
feat(rust): return socket address when starting a transport listener

To reduce boilerplate in tests to get a random port to bind to,
now the router's `bind` method return the socket address that was used.

This can be useful when binding to port 0.

adrianbenavides commented:
I've added the changes suggested by @twittner to remove the "random port" logic altogether.

SanjoDeundiak commented:
How do we manage to collide picking random port from 10000 to 65535?
P.S. Can we reduce the scope of this PR to fixing tests only? Would be glad to discuss macro update in separate PR

adrianbenavides committed:
chore(rust): reverts changes related to the test macro

adrianbenavides commented:
> How do we manage to collide picking random port from 10000 to 65535?

In my case, I had some os processes running on ports 19455, 42849 and 50586. I don't know if a similar case could happen in CI, but knowing what ports are available is out of our control I guess, so it should be best if we can assign a port automatically by the os. Also if tests were to be run in parallel, chances would definitely be there, although this fix won't solve the problem toralf outlined here: https://github.com/build-trust/ockam/issues/2730.

> P.S. Can we reduce the scope of this PR to fixing tests only? Would be glad to discuss macro update in separate PR

Sure, done!

twittner commented on implementations/rust/ockam/ockam_transport_tcp/tests/send_receive.rs:
```diff
@@ -1,33 +1,27 @@
-        transport.listen(bind_address).await?;
+        listener_address = transport.listen("127.0.0.1:0").await?;
         ctx.start_worker("echoer", Echoer).await?;
     };
```
How about removing the `let _listener` block and initialising `listener_address` directly? E.g.

```rust
let transport = TcpTransport::create(ctx).await?;
let listener_address = transport.listen("127.0.0.1:0").await?;
ctx.start_worker("echoer", Echoer).await?;
```

adrianbenavides replied:
Absolutely, thanks!

twittner commented on implementations/rust/ockam/ockam_transport_tcp/tests/send_receive.rs:
```diff
@@ -55,33 +50,27 @@ impl Worker for Echoer {
+    let tx_address = transport.connect(&listener_address).await?;
 
-    let r = route![(TCP, format!("localhost:{}", rand_port)), "echoer"];
+    let r = route![(TCP, listener_address.clone()), "echoer"];
```
`SocketAddr` implements `Copy`, so in theory the clone call is not needed.

adrianbenavides replied:
It expects a `String` there, and also a few lines above that (when calling `transport.connect(&listener_address)`). It looks like cloning that string is the less painful option.

twittner replied:
Ah sorry! You are right of course, I missed that you had it converted to a string again in line 56.

twittner reviewed (dismissed):
:+1:

SanjoDeundiak commented on implementations/rust/ockam/ockam_transport_tcp/src/transport.rs:
```diff
@@ -1,8 +1,11 @@
-use crate::{parse_socket_addr, TcpOutletListenWorker, TcpRouter, TcpRouterHandle};
+use std::net::SocketAddr;
```
The right import would be 
```
use ockam_core::compat::net::SocketAddr;
```
@antoinevg could you please share your recommendations on imports?

adrianbenavides replied:
I've replaced the `std` imports with the `ockam_core::compat` ones.

SanjoDeundiak reviewed (dismissed):
Looks great, one fix is required though

adrianbenavides committed:
refactor(rust): code block and imports

mrinalwadhwa reviewed (changes requested):
@adrianbenavides It looks like `commitlint` [ignores](https://github.com/conventional-changelog/commitlint/blob/914782aad70d353baf4d9fbbf9824c0211241484/%40commitlint/is-ignored/src/is-ignored.test.ts#L96) commits of the form:

```
Revert "test(rust): fix flaky transport tests"
```

That was surprising to me. We should add some overriding lints tests for it.

Since in out workflow commits get rebased and moved around, its better to avoid referring to older commits and instead just describe the change that is in a patch/commit.

adrianbenavides commented:
> @adrianbenavides It looks like `commitlint` [ignores](https://github.com/conventional-changelog/commitlint/blob/914782aad70d353baf4d9fbbf9824c0211241484/%40commitlint/is-ignored/src/is-ignored.test.ts#L96) commits of the form:
> 
> ```
> Revert "test(rust): fix flaky transport tests"
> ```
> 
> That was surprising to me. We should add some overriding lints tests for it.
> 
> Since in out workflow commits get rebased and moved around, its better to avoid referring to older commits and instead just describe the change that is in a patch/commit.

Yeah, I didn't think about changing the auto-generated message after the revert. I've changed it now!
